### DISCUSSION TITLE
Setting BCs in an arbitrary box within a fracture network

### DISCUSSION THREAD

**Discussion by rafaelmarch3** [pmgbergen/porepy#559] [Q&A]
Hello all, 

I hope you're all well. We have a 3D fracture network, and would like to upscale permeability for different boxes within this network. The goal is to determine REV and upscale into single or dual porosity models. 

I was wondering if there's any utility function to set pressure/flow BCs in the boundaries of a given arbitrary box or to cut the fracture network into a smaller model? 

Thank you!
Rafael March.

**Comment by keileg** [comment]
Hi Rafael,

This should be fully doable. It is useful to think of the problem as first restricting the fracture network to a specific domain, and next to solve a flow problem on that domain.

The `domain `in `FractureNetwork3d`'s init can be specified as set of bounding planes, specified as a list of `numpy `arrays. These should form a closed volume (I don't think you will get an error message if you forget a side, so be careful), and can define more or less any polyhedra. Fractures that are outside the imposed domain will be dropped and fractures that cross the domain boundary will be cropped. You may want to use the function `impose_external_boundary()` in `FractureNetwork3d `directly (usually this called automatically when you run the `mesh()` function) to be more in control of what happens.

 Two warnings:
1. Although the geometry computations that take place under the hood have become quite robust, the usual caveats about complex geometries requiring many cells, and possibly tuning of tolerances, apply.
2. When the boundary is imposed, and fractures outside the domain are dropped, you loose the 1-1 relation between the ordering of fractures when you feed them to `FractureNetwork3d`, and the ordering of 2d fracture grids. Thus, you need to be careful if you want to specify different permeabilities on different fractures or similar. The function that drops fractures will return a list with the indices of preserved fractures, but this is not the best documented part of the code. Let me know if you have questions.

The above should give you a standard `GridBucket` that covers only the specified part of the domain. To set boundary conditions on this domain, you will need to identify faces (both in the 3d grid, and 2d fracture grids) that touch the relevant part of the boundary. Since you know the bounding planes of the domain, this should be doable.

Final comment: I have several minor fixes to the meshing in a branch that has not yet been merged. Let me know if you run into meshing issues for complex geometries, then I and will try to speed up merging in the hope that this will help. 

Let me know if this makes sense to you.

**Comment by sanek151** [comment]
> Hello all,
> 
> I hope you're all well. We have a 3D fracture network, and would like to upscale permeability for different boxes within this network. The goal is to determine REV and upscale into single or dual porosity models.
> 
> I was wondering if there's any utility function to set pressure/flow BCs in the boundaries of a given arbitrary box or to cut the fracture network into a smaller model?
> 
> Thank you! Rafael March.

241sdf
